File: java/src/main/java/io/binac/leetcode/TextJustification.java
package io.binac.leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * Given an array of words and a width maxWidth, format the text such that each line has exactly maxWidth characters and is fully (left and right) justified.
 * <p>
 * <p>You should pack your words in a greedy approach; that is, pack as many words as you can in each line. Pad extra spaces ' ' when necessary so that each line has exactly maxWidth characters.
 * <p>
 * <p>Extra spaces between words should be distributed as evenly as possible. If the number of spaces on a line do not divide evenly between words, the empty slots on the left will be assigned more spaces than the slots on the right.
 * <p>
 * <p>For the last line of text, it should be left justified and no extra space is inserted between words.
 * <p>
 * <p>Note:
 * <p>
 * <p>A word is defined as a character sequence consisting of non-space characters only.
 * <p>Each word's length is guaranteed to be greater than 0 and not exceed maxWidth.
 * <p>The input array words contains at least one word.
 * <p>
 * <p>Example 1:
 * <blockquote><pre>
 *     Input:
 *     words = ["This", "is", "an", "example", "of", "text", "justification."]
 *     maxWidth = 16
 *     Output:
 *     [
 *        "This    is    an",
 *        "example  of text",
 *        "justification.  "
 *     ]
 * </blockquote></pre>
 * Example 2:
 * <blockquote><pre>
 *     Input:
 *     words = ["What","must","be","acknowledgment","shall","be"]
 *     maxWidth = 16
 *     Output:
 *     [
 *       "What   must   be",
 *       "acknowledgment  ",
 *       "shall be        "
 *     ]
 *     Explanation: Note that the last line is "shall be    " instead of "shall     be",
 *                  because the last line must be left-justified instead of fully-justified.
 *                  Note that the second line is also left-justified becase it contains only one word.
 * </blockquote></pre>
 * Example 3:
 * <blockquote><pre>
 *     Input:
 *     words = ["Science","is","what","we","understand","well","enough","to","explain",
 *              "to","a","computer.","Art","is","everything","else","we","do"]
 *     maxWidth = 20
 *     Output:
 *     [
 *       "Science  is  what we",
 *       "understand      well",
 *       "enough to explain to",
 *       "a  computer.  Art is",
 *       "everything  else  we",
 *       "do                  "
 *     ]
 * </blockquote></pre>
 */
public class TextJustification {
    public static class Solution1 {
        private void appendSpace(StringBuilder sb, int count) {
            for (; count > 0; --count)
                sb.append(' ');
        }

        public List<String> fullJustify(String[] words, int maxWidth) {
            if (maxWidth <= 0)
                throw new IllegalArgumentException("maxWidth must be positive");
            if (words.length == 0)
                throw new IllegalArgumentException("words must contain at least one word");
            List<String> result = new ArrayList<>();
            int i = 0;
            while (i < words.length) {
                int start = i, len = words[i].length(), count = 1;
                while (i + 1 < words.length && len + words[i + 1].length() + count <= maxWidth) {
                    len += words[++i].length();
                    ++count;
                }
                StringBuilder sb = new StringBuilder();
                if (i < words.length - 1) {
                    int space = maxWidth - len;
                    final int quotient = count == 1 ? space : space / (count - 1);
                    int remainder = count == 1 ? 0 : space % (count - 1);
                    sb.append(words[start]);
                    for (int j = start + 1; j <= i; ++j) {
                        appendSpace(sb, quotient);
                        if (remainder > 0) {
                            sb.append(' ');
                            --remainder;
                        }
                        sb.append(words[j]);
                    }
                } else {
                    sb.append(words[start]);
                    for (int j = start + 1; j <= i; ++j) {
                        sb.append(' ');
                        sb.append(words[j]);
                    }
                }
                appendSpace(sb, maxWidth - sb.length());
                result.add(sb.toString());
                ++i;
            }
            return result;
        }
    }
}

File: python3/leetcodepy/text_justification.py
"""
Given an array of strings words and a width maxWidth, format the text such that each line has exactly maxWidth characters and is fully (left and right) justified.

You should pack your words in a greedy approach; that is, pack as many words as you can in each line. Pad extra spaces ' ' when necessary so that each line has exactly maxWidth characters.

Extra spaces between words should be distributed as evenly as possible. If the number of spaces on a line does not divide evenly between words, the empty slots on the left will be assigned more spaces than the slots on the right.

For the last line of text, it should be left-justified and no extra space is inserted between words.

Note:

A word is defined as a character sequence consisting of non-space characters only.
Each word's length is guaranteed to be greater than 0 and not exceed maxWidth.
The input array words contains at least one word.


Example 1:

Input: words = ["This", "is", "an", "example", "of", "text", "justification."], maxWidth = 16
Output:
[
   "This    is    an",
   "example  of text",
   "justification.  "
]

Example 2:

Input: words = ["What","must","be","acknowledgment","shall","be"], maxWidth = 16
Output:
[
  "What   must   be",
  "acknowledgment  ",
  "shall be        "
]
Explanation: Note that the last line is "shall be    " instead of "shall     be", because the last line must be left-justified instead of fully-justified.
Note that the second line is also left-justified becase it contains only one word.

Example 3:

Input: words = ["Science","is","what","we","understand","well","enough","to","explain","to","a","computer.","Art","is","everything","else","we","do"], maxWidth = 20
Output:
[
  "Science  is  what we",
  "understand      well",
  "enough to explain to",
  "a  computer.  Art is",
  "everything  else  we",
  "do                  "
]


Constraints:

1 <= words.length <= 300
1 <= words[i].length <= 20
words[i] consists of only English letters and symbols.
1 <= maxWidth <= 100
words[i].length <= maxWidth
"""
from typing import List


class Solution1:
    def fullJustify(self, words: List[str], maxWidth: int) -> List[str]:
        result = []
        i = 0
        size = len(words)
        while i < size:
            j = i + 1
            length = len(words[i])
            while j < size and length + 1 + len(words[j]) <= maxWidth:
                length += 1 + len(words[j])
                j += 1
            count = j - i
            line = words[i]
            if j < size and count > 1:
                space = maxWidth - length + count - 1
                space_each = space // (count - 1)
                space_extra = space % (count - 1)
                for k in range(i + 1, j):
                    line += ' ' * space_each
                    if space_extra > 0:
                        line += ' '
                        space_extra -= 1
                    line += words[k]
            else:
                for k in range(i + 1, j):
                    line += ' ' + words[k]
                line += ' ' * (maxWidth - len(line))
            result.append(line)
            i = j
        return result
